pkg/plugins/diff/delta_diff_server/src/main.rs

```diff
@@ -0,0 +1,259 @@
+impl TableDiffer for DifferService {
+    /*
+    table_diff will compare two states of a given table over two branches in the following way:
+    1. If the destination table (the one which the source table branch is compared to) isn't found
+       then it's assumed that the table has been created, and all of its Delta Log will be returned as an answer.
+    2. If the source table (the one which the user asked to compare to some destination table) isn't found
+       then it's assumed that the table has been dropped.
+    3. If both tables are found:
+        A. If the right table's version is bigger than the left table's version: add all versions of the right tables that are bigger than
+           the left table to the returned answer.
+        B. Mark the latest common version as the start of the comparison algorithm.
+        C. Mark the earliest common version as the end of the comparison algorithm.
+        D. Compare the versions (commits) of the tables starting from the version found on step 'B':
+            a. If the "commit info"s (Delta's commits) are different, add the right table's commit info to the result.
+            b. Else, a common commit was found and the diff is complete.
+     */
+    async fn table_diff(&self, request: Request<DiffRequest>) -> Result<Response<DiffResponse>, Status> {
+        let r = request.into_inner();
+        let s3_gateway_config_req: GatewayConfig = r.gateway_config.expect("Missing S3 compatible gateway configurations");
+        let diff_props: DiffProps = r.props.expect("Missing diff properties");
+        let left_table_path: TablePath = diff_props.left_table_path.expect("Missing left table's path");
+        let right_table_path: TablePath = diff_props.right_table_path.expect("Missing right table's path");
+
+        let s3_config_map: HashMap<String, String> = utils::construct_storage_config(s3_gateway_config_req);
+        let left_table_res =
+            delta_ops::get_delta_table(&s3_config_map, &diff_props.repo, &left_table_path);
+        let right_table_res =
+            delta_ops::get_delta_table(&s3_config_map, &diff_props.repo, &right_table_path);
+
+        // Get the base path, or return a default table path (with all fields initialized to their default type values)
+        let base_path = diff_props.base_table_path.unwrap_or(TablePath{ r#ref: "".to_string(), path: "".to_string() });
+        let base_table_res =
+            delta_ops::get_delta_table(&s3_config_map, &diff_props.repo, &base_path);
+
+        let (left_table_res, right_table_res, base_table_res) =
+            join!(left_table_res, right_table_res, base_table_res);
+
+        if right_table_res.is_err() || left_table_res.is_err() {
+            return handle_table_err(left_table_res, right_table_res, base_table_res).await;
+        }
+
+        let mut right_table = right_table_res?;
+        let mut left_table= left_table_res?;
+        return run_diff(&mut left_table, &mut right_table).await;
```

Comment:
A suggestion - 
```suggestion
        let diffType = getTableDiffType(right_table_res, left_table_res)
        match diffType {
            (DiffType::Created) => {}
            (DiffType::Dropped) => {}
            (DiffType::Changed) => {}   / this is the default
        }
```
and encapsulate the logic in handle error + in lines 80-82 in the invoked functions.

What's a bit confusing in the current implementation is that the outcome of handle_table_err can be a valid table diff. but it is up to you

Reply:
you got it